### Data Augmentation using ImageDataGenerator in Keras

In this notebook, I will seek to build on previously built model's by exploiting Keras' built in data augmentation functionality. The ```ImageDataGenerator``` class provides an infinite stream of manipulated images. It takes in the training data as normal and outputs images that have been transformed, rotated, shifted, and flipped. 

These new images provide the basis for a new training set for the neural network, with the hope that introducing some new changes will help the network discover new patterns in the data it could not with the fixed training set. 

In [1]:
%run __initremote__.py

Using TensorFlow backend.


x_train shape: (50000, 32, 32, 3)
50000 train samples
10000 test samples


In [2]:
from keras.models import load_model

In [8]:
model = load_model('60_epoch_adam.h5')

In [11]:
early_stop = keras.callbacks.EarlyStopping(monitor='val_acc', 
                                           min_delta=0, 
                                           patience=10, 
                                           verbose=0, 
                                           mode='auto')

In [10]:
opt = keras.optimizers.adam(lr=0.0001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

In [12]:
datagen = ImageDataGenerator(rotation_range=0, 
                             width_shift_range=0.1, 
                             height_shift_range=0.1, 
                             horizontal_flip=True)

datagen.fit(x_train)

model.fit_generator(datagen.flow(x_train, y_train, batch_size=32),
                   epochs=100,
                   validation_data=(x_test, y_test),
                   callbacks=[early_stop],
                   workers=-1)

Epoch 1/100
1563/1563 [==============================] - 46s 30ms/step - loss: 0.7741 - acc: 0.7550 - val_loss: 0.5788 - val_acc: 0.8067
Epoch 2/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.7293 - acc: 0.7562 - val_loss: 0.5719 - val_acc: 0.8077
Epoch 3/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.7093 - acc: 0.7614 - val_loss: 0.5646 - val_acc: 0.8086
Epoch 4/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.6967 - acc: 0.7641 - val_loss: 0.5637 - val_acc: 0.8102
Epoch 5/100
1563/1563 [==============================] - 47s 30ms/step - loss: 0.6833 - acc: 0.7656 - val_loss: 0.5620 - val_acc: 0.8129
Epoch 6/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.6850 - acc: 0.7679 - val_loss: 0.5670 - val_acc: 0.8093
Epoch 7/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.6765 - acc: 0.7701 - val_loss: 0.5562 - val_acc: 0.8143
Epoch 8/100
1563/1563 [==================

1563/1563 [==============================] - 46s 29ms/step - loss: 0.5310 - acc: 0.8151 - val_loss: 0.4771 - val_acc: 0.8387
Epoch 64/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.5361 - acc: 0.8152 - val_loss: 0.4878 - val_acc: 0.8342
Epoch 65/100
1563/1563 [==============================] - 46s 30ms/step - loss: 0.5291 - acc: 0.8157 - val_loss: 0.4834 - val_acc: 0.8366
Epoch 66/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.5317 - acc: 0.8153 - val_loss: 0.4845 - val_acc: 0.8369
Epoch 67/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.5291 - acc: 0.8152 - val_loss: 0.4791 - val_acc: 0.8380
Epoch 68/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.5242 - acc: 0.8189 - val_loss: 0.4823 - val_acc: 0.8371
Epoch 69/100
1563/1563 [==============================] - 47s 30ms/step - loss: 0.5299 - acc: 0.8178 - val_loss: 0.4772 - val_acc: 0.8401
Epoch 70/100
1563/1563 [=======================

After 100 epochs, the model is still learning while using the ```datagen.flow``` method. Let's run it for another 100 epochs and see if it can get a bit smarter. At this point the model has been training for approximately 80 minutes, not including prior learning.

In [13]:
model.fit_generator(datagen.flow(x_train, y_train, batch_size=32),
                   epochs=100,
                   validation_data=(x_test, y_test),
                   callbacks=[early_stop],
                   workers=-1)

Epoch 1/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.4977 - acc: 0.8276 - val_loss: 0.4594 - val_acc: 0.8470
Epoch 2/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.4913 - acc: 0.8301 - val_loss: 0.4615 - val_acc: 0.8476
Epoch 3/100
1563/1563 [==============================] - 46s 30ms/step - loss: 0.4863 - acc: 0.8311 - val_loss: 0.4654 - val_acc: 0.8446
Epoch 4/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.4938 - acc: 0.8292 - val_loss: 0.4645 - val_acc: 0.8439
Epoch 5/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.4865 - acc: 0.8306 - val_loss: 0.4589 - val_acc: 0.8469
Epoch 6/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.4863 - acc: 0.8311 - val_loss: 0.4590 - val_acc: 0.8463
Epoch 7/100
1563/1563 [==============================] - 46s 29ms/step - loss: 0.4840 - acc: 0.8322 - val_loss: 0.4586 - val_acc: 0.8452
Epoch 8/100
1563/1563 [==================

In [14]:
model.evaluate(x_test, y_test)

10000/10000 [==============================] - 1s 146us/step


[0.45529800367355344, 0.84589999999999999]

In [15]:
model.save('datagen_115.h5')

Additional training was not able to significantly help the model learn. Currently, it tests at a 84.6% accuracy, with a test error rate of 15.4%. This result is starting to be comparative to other benchmark models for CIFAR-10, referenced on the cs.toronto.edu readme page. 

From cs.toronto.edu: "You can find some baseline replicable results on this dataset on the project page for cuda-convnet. These results were obtained with a convolutional neural network. Briefly, they are 18% test error without data augmentation and 11% with." See http://code.google.com/p/cuda-convnet/

Prior to data augmentation, the test error rate was just about 18.5%, again comparative to the baseline results of work done elsewhere. 